### PR TITLE
Apply p3c pmd plugin/rules

### PR DESCRIPTION
**Pull request by blindpirate**
### Ⅰ. Describe what this PR did

This PR applies alibaba p3c pmd plugin and rules.

### Ⅱ. Does this pull request fix one issue?

No

### Ⅲ. Why don't you add test cases (unit test/integration test)? 

This PR doesn't change any production code.

### Ⅳ. Describe how to verify it

```
mvn clean install -DskipTests
```

### CODE REVIEW

**Comment by slievrly** [comment]
@blindpirate good job

**Comment by slievrly** [comment]
http://maven.apache.org/plugins/maven-pmd-plugin/

**Comment by slievrly** [comment]
@blindpirate When maven execution includes `mvn verify`，it work?

**Comment by blindpirate** [comment]
@jovany-wang Nice catch! I didn't notice that `install` in `.travis.yml` was skipped. I've added `verify` to `.travis.yml`'s `script`.

@slievrly Yes, `fescar` has no violations of priority 1 (but has a few violations of lower priority).

File: .travis.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ cache:
 install: true
 
 script:
-  - mvn clean package
+  - mvn clean package verify
 
 after_success:
   - bash <(curl -s https://codecov.io/bash)
```
Comment (jovany-wang) [inline .travis.yml]: Add a new line at the end of this file.

Comment (blindpirate) [reply .travis.yml]: Done.

**Comment by blindpirate** [comment]
Thanks @jovany-wang @slievrly !

**Comment by lovepoem** [comment]
image

Is this right ? 
 How to verify that your P3C check script is working ？

**Comment by blindpirate** [comment]
@lovepoem Remove this line https://github.com/alibaba/fescar/blob/develop/pom.xml#L380 `<minimumPriority>1</minimumPriority>`, then run `mvn clean package verify` you can see the failures of violation in priority 2 and 3.

The `@author` failure in screenshot is of priority 3.

**Comment by slievrly** [comment]
@blindpirate  I read "does-alibaba-follow-their-own-guidelines" last week，so I made code changes to fescar. I believe there is no "default" level specification problem. Can we help you prove it with your tools?

**Comment by blindpirate** [comment]
@slievrly With current `develop` branch, if I remove `<minimumPriority>1</minimumPriority>` in pom.xml, then run `mvn clean package verify`, I got this on my local machine:

```
[INFO] --- maven-pmd-plugin:3.8:check (default) @ fescar-common ---
[INFO] PMD Failure: com.alibaba.fescar.common.exception.NotSupportYetException:22 Rule:ClassMustHaveAuthorRule Priority:3 【NotSupportYetException】注释缺少@author信息.
[INFO] PMD Failure: com.alibaba.fescar.common.exception.ShouldNeverHappenException:22 Rule:ClassMustHaveAuthorRule Priority:3 【ShouldNeverHappenException】注释缺少@author信息.
[INFO] PMD Failure: com.alibaba.fescar.common.thread.RejectedPolicies:28 Rule:ClassMustHaveAuthorRule Priority:3 【RejectedPolicies】注释缺少@author信息.
[INFO] PMD Failure: com.alibaba.fescar.common.util.BlobUtils:30 Rule:ClassMustHaveAuthorRule Priority:3 【BlobUtils】注释缺少@author信息.
[INFO] PMD Failure: com.alibaba.fescar.common.util.StringUtils:30 Rule:ClassMustHaveAuthorRule Priority:3 【StringUtils】注释缺少@author信息.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Alibaba fescar 0.3.1 0.3.1 ......................... SUCCESS [  1.518 s]
[INFO] fescar-common 0.3.1 ................................ FAILURE [  4.860 s]
[INFO] fescar-config 0.3.1 ................................ SKIPPED
[INFO] fescar-discovery 0.3.1 ............................. SKIPPED
[INFO] fescar-core 0.3.1 .................................. SKIPPED
[INFO] fescar-server 0.3.1 ................................ SKIPPED
[INFO] fescar-distribution 0.3.1 .......................... SKIPPED
[INFO] fescar-rm-datasource 0.3.1 ......................... SKIPPED
[INFO] fescar-tm 0.3.1 .................................... SKIPPED
[INFO] fescar-spring 0.3.1 ................................ SKIPPED
[INFO] fescar-dubbo 0.3.1 ................................. SKIPPED
[INFO] fescar-test 0.3.1 .................................. SKIPPED
[INFO] fescar-dubbo-alibaba 0.3.1 0.3.1 ................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 6.879 s
[INFO] Finished at: 2019-03-17T20:43:59+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.8:check (default) on project fescar-common: You have 5 PMD violations. For more details see: /Users/zhb/Projects/tmp/fescar/common/target/pmd.xml -> [Help 1]
```

Or am I misunderstanding you?

**Comment by blindpirate** [comment]
@slievrly With prority 2 I got:

```
[INFO] --- maven-pmd-plugin:3.8:check (default) @ fescar-config ---
[INFO] PMD Failure: com.alibaba.fescar.config.ZKConfiguration:56 Rule:ConstantFieldShouldBeUpperCaseRule Priority:2 常量【executor】命名应全部大写并以下划线分隔.
[INFO] PMD Failure: com.alibaba.fescar.config.ZKConfiguration:56 Rule:ThreadShouldSetNameRule Priority:2 要使用带有ThreadFactory参数的ThreadPoolExecutor构造方法哦，这样你就可以方便的设置线程名字啦。.
```

**Comment by slievrly** [comment]
@blindpirate Is it convenient to contact you by dingTalk?

**Comment by blindpirate** [comment]
@slievrly 我没有钉钉账号，如果你愿意的话我们可以直接中文这边聊天。

**Comment by slievrly** [comment]
@blindpirate 好的，可以。

**Comment by slievrly** [comment]
image
@blindpirate 我想先看下这个是否消除了。

**Comment by slievrly** [comment]
@blindpirate 我估计对你表格的“default” 理解有歧义，我先按照prority=2  的级别逐步改下。

**Comment by blindpirate** [comment]
这个`5 PMD violations`未必是整个项目的violations数量， 因为只要任何一个子项目报错`X PMD violations`，整个构建就进行不下去了，所以这个数量其实没有很大意义。`5 PMD violations`是上周的某个时间点我运行`verify`时候得到的结果，当时具体的violation是什么我没有保存，所以实在抱歉我不能知道是否消除了。

**Comment by slievrly** [comment]
@blindpirate 我是不是可以这样理解 default 是指去掉了<minimumPriority>1</minimumPriority>，但是因为构建到一个模块出错的时候就停了，也就是说只是一个模块的。

**Comment by blindpirate** [comment]
我对default的理解是，不配置任何的优先级运行verify。这会把所有的violations (1/2/3)全部打印出来。后来有人告诉我，只有【强制】（优先级1）的violation是需要处理的，因此我把`minimumPriority`设置成了1.

**Comment by slievrly** [comment]
@blindpirate 好的，我明白了。我们对于 1 级别的应该没什么问题，我们会设置成 2 改一下。

**Comment by blindpirate** [comment]
:-)

有问题随时联系。

**Comment by slievrly** [comment]
@blindpirate 多谢支持。

**Pull request merged**